File: programs/tokenizer/src/state/dividend_distribution.rs
use pinocchio::Address;

use super::DIVIDEND_DISTRIBUTION_SEED;

/// Dividend distribution for an asset. Holds escrow with deposited stablecoins
/// and tracks total_amount / total_shares for pro-rata payout calculation.
/// PDA: ["dividend_distribution", asset.key(), &epoch.to_le_bytes()]
#[repr(C)]
pub struct DividendDistribution {
    // Discriminator
    pub account_key: u8,              // AccountKey::DividendDistribution
    pub version: u8,

    // Identity
    pub asset: [u8; 32],             // Parent Asset PDA
    pub epoch: u32,                   // Distribution epoch (matches asset.dividend_epoch at creation)

    // Distribution config
    pub accepted_mint: [u8; 32],     // Which stablecoin
    pub total_amount: u64,            // Total stablecoins deposited for distribution
    pub total_shares: u64,            // Snapshot of asset.minted_shares at creation time
    pub shares_claimed: u64,          // Running total of shares claimed so far

    // Escrow
    pub escrow: [u8; 32],            // Escrow token account address

    // Timestamps
    pub created_at: i64,

    // PDA bumps
    pub bump: u8,
    pub escrow_bump: u8,
}

impl DividendDistribution {
    pub const LEN: usize = core::mem::size_of::<Self>();

    #[inline(always)]
    pub unsafe fn load(data: &[u8]) -> &Self {
        &*(data.as_ptr() as *const Self)
    }

    #[inline(always)]
    pub unsafe fn load_mut(data: &mut [u8]) -> &mut Self {
        &mut *(data.as_mut_ptr() as *mut Self)
    }

    pub fn derive_pda(
        asset: &[u8; 32],
        epoch: u32,
        program_id: &Address,
    ) -> (Address, u8) {
        Address::find_program_address(
            &[DIVIDEND_DISTRIBUTION_SEED, asset, &epoch.to_le_bytes()],
            program_id,
        )
    }
}

File: sdks/tokenizer-sdk/src/accounts/dividendDistribution.ts
/**
 * DividendDistribution account deserializer.
 *
 * #[repr(C)] layout (144 bytes):
 *   0: account_key (u8)
 *   1: version (u8)
 *   2: asset ([u8;32])
 *  36: epoch (u32)              — padded from 34
 *  40: accepted_mint ([u8;32])
 *  72: total_amount (u64)
 *  80: total_shares (u64)
 *  88: shares_claimed (u64)
 *  96: escrow ([u8;32])
 * 128: created_at (i64)
 * 136: bump (u8)
 * 137: escrow_bump (u8)
 * 138: rent_payer ([u8;32])
 */
import { type Address, getStructDecoder, transformDecoder } from "gill";
import { AccountKey } from "../constants.js";
import { addr, i64d, pad, u8d, u32d, u64d } from "./decode.js";

export const DIVIDEND_DISTRIBUTION_SIZE = 176;
export const DIVIDEND_DISTRIBUTION_OFFSET_ASSET = 2;

export interface DividendDistribution {
  accountKey: number;
  version: number;
  asset: Address;
  epoch: number;
  acceptedMint: Address;
  totalAmount: bigint;
  totalShares: bigint;
  sharesClaimed: bigint;
  escrow: Address;
  createdAt: bigint;
  bump: number;
  escrowBump: number;
  rentPayer: Address;
}

const rawDecoder = getStructDecoder([
  ["accountKey", u8d],
  ["version", u8d],
  ["asset", addr],
  ["_p0", pad(2)],
  ["epoch", u32d],
  ["acceptedMint", addr],
  ["totalAmount", u64d],
  ["totalShares", u64d],
  ["sharesClaimed", u64d],
  ["escrow", addr],
  ["createdAt", i64d],
  ["bump", u8d],
  ["escrowBump", u8d],
  ["rentPayer", addr],
  ["_p1", pad(6)],
]);

export const dividendDistributionDecoder = transformDecoder(
  rawDecoder,
  ({ _p0, _p1, ...rest }) => rest,
);

export function decodeDividendDistribution(data: Uint8Array): DividendDistribution {
  if (data.length < DIVIDEND_DISTRIBUTION_SIZE) {
    throw new Error(
      `DividendDistribution: expected ${DIVIDEND_DISTRIBUTION_SIZE} bytes, got ${data.length}`,
    );
  }
  const accountKey = data[0];
  if (accountKey !== AccountKey.DividendDistribution) {
    throw new Error(`DividendDistribution: invalid account key ${accountKey}`);
  }
  return dividendDistributionDecoder.decode(data);
}
